zookeeper/core/factory.py

```diff
@@ -0,0 +1,106 @@
+import functools
+import inspect
+from typing import Type
+
+from zookeeper.core import utils
+from zookeeper.core.component import component
+from zookeeper.core.factory_registry import FACTORY_REGISTRY
+
+
+# A sentinel class/object for missing default values.
+class Missing:
+    def __repr__(self):
+        return f"<missing>"
+
+
+missing = Missing()
+
+
+def _wrap_build(factory_cls: Type) -> None:
+    """
+    Every @factory has a `build()` method, which we wrap so that `build()` is
+    only called once (lazily) and the value is cached.
+    """
+    fn = factory_cls.build
+
+    @functools.wraps(fn)
+    def wrapped_fn(factory_instance):
+        if factory_instance.__component_factory_value__ is missing:
+            result = fn(factory_instance)
+            if not utils.type_check(
+                result, factory_cls.__component_factory_return_type__
+            ):
+                raise TypeError(
+                    f"@factory '{factory_cls}' has a `build()` method annotated with "
+                    f"return type {factory_cls.__component_factory_return_type__}, "
+                    f"which is not satisfied by the return value {result}."
+                )
+            factory_instance.__component_factory_value__ = result
+        return factory_instance.__component_factory_value__
+
+    factory_cls.build = wrapped_fn
+
+
+def _wrap_str_repr(factory_cls: Type) -> None:
+    str_fn = factory_cls.__str__
+    repr_fn = factory_cls.__repr__
+
+    @functools.wraps(str_fn)
+    def wrapped_str(factory_instance):
+        result = str_fn(factory_instance)
+        return result.replace("<Unconfigured component ", "<Unconfigured factory ")
+
+    @functools.wraps(repr_fn)
+    def wrapped_repr(factory_instance):
+        result = repr_fn(factory_instance)
+        return result.replace("<Unconfigured component ", "<Unconfigured factory ")
+
+    factory_cls.__str__ = wrapped_str
+    factory_cls.__repr__ = wrapped_repr
+
+
+def factory(cls: Type):
+    cls = component(cls)
+
+    try:
+        signature = inspect.signature(cls.build)
+        params = signature.parameters
+        if (
+            len(params) != 1
+            or "self" not in params
```

Comment:
I'm going to leave this for now because this behaviour mirrors the existing behaviour in `zookeeper/core/component.py` but we can revisit this later after this PR is merged.